bitbots_blackboard/bitbots_blackboard/capsules/costmap_capsule.py

```diff
@@ -0,0 +1,390 @@
+
+if TYPE_CHECKING:
+    from bitbots_blackboard.blackboard import BodyBlackboard
+
+import matplotlib.pyplot as plt
+import numpy as np
+import ros2_numpy
+import tf2_ros as tf2
+from bitbots_utils.utils import (get_parameter_dict,
+                                 get_parameters_from_other_node)
+from geometry_msgs.msg import (Point, Pose, PoseStamped,
+                               PoseWithCovarianceStamped, Quaternion,
+                               TransformStamped, TwistStamped,
+                               TwistWithCovarianceStamped)
+from nav_msgs.msg import MapMetaData, OccupancyGrid
+from PIL import Image, ImageDraw
+from rclpy.clock import ClockType
+from rclpy.duration import Duration
+from rclpy.time import Time
+from scipy.interpolate import griddata
+from scipy.ndimage import gaussian_filter
+from soccer_vision_3d_msgs.msg import RobotArray, Robot
+from std_msgs.msg import Header
+from std_srvs.srv import Trigger
+from tf2_geometry_msgs import PointStamped
+from tf_transformations import euler_from_quaternion, quaternion_from_euler
+
+class CostmapCapsule:
+    def __init__(self, blackboard: "BodyBlackboard"):
+        self.blackboard = blackboard
+        self.body_config = get_parameter_dict(self._blackboard.node, "body")
+
+
+        self.tf_buffer = tf2.Buffer(cache_time=Duration(seconds=30))
+        self.tf_listener = tf2.TransformListener(self.tf_buffer, self._blackboard.node)
+
+        self.base_footprint_frame: str = self._blackboard.node.get_parameter('base_footprint_frame').value
+
+        parameters = get_parameters_from_other_node(
+            self._blackboard.node,
+            "/parameter_blackboard",
+            ["field_length", "field_width", "goal_width"])
+        self.field_length: float = parameters["field_length"]
+        self.field_width: float = parameters["field_width"]
+        self.goal_width: float = parameters["goal_width"]
+        self.map_margin: float = self._blackboard.node.get_parameter('body.map_margin').value
+        self.obstacle_costmap_smoothing_sigma: float = self._blackboard.node.get_parameter(
+            'body.obstacle_costmap_smoothing_sigma').value
+        self.obstacle_cost: float = self._blackboard.node.get_parameter('body.obstacle_cost').value
+
+
+        # Publisher for visualization in RViZ
+        self.costmap_publisher = self._blackboard.node.create_publisher(OccupancyGrid, 'debug/costmap', 1)
+
+        self.base_costmap: Optional[np.ndarray] = None  # generated once in constructor field features
+        self.costmap: Optional[np.ndarray] = None  # updated on the fly based on the base_costmap
+        self.gradient_map: Optional[np.ndarray] = None  # global heading map (static) only dependent on field structure
+
+        # Calculates the base costmap and gradient map based on it
+        self.calc_base_costmap()
+        self.calc_gradients()
+
+
+    ############
+    # Obstacle #
+    ############
+
+    def robot_obstacle_callback(self, msg: RobotArray):
```

Comment:
Maybe just robot_callback

Reply:
I can refactor that. Where is robot_obstacle_callback supposed to be called? When looking for it, this method is never used.